nestadia/src/ppu/mod.rs

```diff
@@ -273,59 +326,123 @@ impl Ppu {
-                if self.ctrl_reg.contains(registers::ControlReg::GENERATE_NMI) {
-                    self.vblank_nmi_set = true;
-                }
-            } else if self.scanline >= 261 {
+            if self.scanline >= 261 {
                 // http://wiki.nesdev.com/w/index.php/PPU_rendering#Pre-render_scanline_.28-1_or_261.29
                 // scanline = -1 is the dummy scanline
                 self.scanline = -1;
 
+                // Skips a cycle on odd frame, only if rendering is enabled
+                if self.is_odd_frame && self.rendering_enabled() {
+                    self.cycle_count += 1
+                };
+
+                self.is_odd_frame = !self.is_odd_frame;
+            }
+
+            // Update the state machine to detect sprite 0
+            match self.sprite_zero_hit_state {
+                SpriteZeroHitState::IsInOam => {
+                    self.sprite_zero_hit_state = SpriteZeroHitState::OnCurrentScanline(false)
+                }
+                SpriteZeroHitState::OnCurrentScanline(false) => {
+                    self.sprite_zero_hit_state = SpriteZeroHitState::Idle
+                }
+                SpriteZeroHitState::OnCurrentScanline(true) => {
+                    self.sprite_zero_hit_state = SpriteZeroHitState::OnCurrentScanline(false)
+                }
+                _ => {}
+            }
+        };
+
+        // Handle sprite 0 delay
+        if let SpriteZeroHitState::Delay(mut x) = self.sprite_zero_hit_state {
+            x -= 1;
+
+            if x == 0 {
+                self.status_reg
+                    .insert(registers::StatusReg::SPRITE_ZERO_HIT);
+                self.sprite_zero_hit_state = SpriteZeroHitState::Idle
+            } else {
+                self.sprite_zero_hit_state = SpriteZeroHitState::Delay(x)
+            }
+        };
+
+        // Pre-render scanline
+        if self.scanline == -1 {
+            if self.cycle_count == 1 {
+                // Reset sprite 0 hit
+                self.status_reg
+                    .remove(registers::StatusReg::SPRITE_ZERO_HIT);
+
+                // Reset sprite overflow flag
+                self.status_reg
+                    .remove(registers::StatusReg::SPRITE_OVERFLOW);
+
                 // VBLANK is done
                 self.status_reg.remove(registers::StatusReg::VBLANK_STARTED);
+            } else if self.cycle_count >= 280
+                && self.cycle_count <= 304
+                && (self.mask_reg.contains(registers::MaskReg::SHOW_BACKGROUND)
+                    || self.mask_reg.contains(registers::MaskReg::SHOW_SPRITES))
+            {
+                self.vram_addr.reset_y(&self.temp_vram_addr);
             }
         };
 
-        self.render_pixel(bus);
-    }
+        self.render_pixel();
 
-    fn render_pixel(&mut self, bus: &mut PpuBus) {
-        use core::convert::TryFrom;
+        // This condition is there to ensure that the VRAM address does not get updated if rendering is turned off
```

Comment:
But it could get updated just above (line 393) though?

Reply:
That was a partial refactor! The RENDER_PIXEL || RENDER_BACKGROUND condition means rendering is enabled, and I've extracted it in a function but forgot to change it right above line 393. I'm fixing it now!